### PR TITLE
[WIP] Implement RFC0035 II: Electric Boogaloo

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

Follow-up for https://github.com/NixOS/nixpkgs/pull/49398

##### TODO:

Should I split this up in two PR's? Part of this is good to go and should just go in the working tree straight away.

- [x] The basic change
- [x] Consistency check
- [x] Check the entire tree is consistent (it is now!)
- [ ] Document the new check
- [x] Docs for the new `hasInfix` library function
- [x] Currently interacts really weirdly with Coq, and I have no idea why.
- [x] Check everything still works.

cc @dtzWill 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Synthetica9** [comment]
[Ofborg's current output](https://gist.github.com/58cc1bd7e85abeb728c72a914425cc22) is the problem I'm having with Coq, and I'm not sure what is even happening, so any help is welcome there

**Comment by Zimmi48** [comment]
cc @vbgl You might have an idea about the Coq-related issue in this PR.

**Comment by Zimmi48** [comment]
@Synthetica9 If I understand correctly, this allows the scheme `name = "ocaml${ocaml.version}-${pname}-${version}"` or `name = "coq${coq.version}-${pname}-${version}"`, etc. Correct?

**Comment by Synthetica9** [comment]
> @Synthetica9 If I understand correctly, this allows the scheme name = "ocaml${ocaml.version}-${pname}-${version}" or name = "coq${coq.version}-${pname}-${version}", etc. Correct?

Correct. The current check is whether both `pname` and `version` (if the attributes exist) appear in `name` in their entirety. So any prefix, postfix, whatever is allowed. This was the most forgiving check I could think of that still meant anything, and it found a few genuine problems (see: most of the commits in this PR).

**Comment by vbgl** [comment]
May I suggest that the issue is in your code rather than in the Coq package descriptions? It seems that the checks are run too early.

Please do remove the failing assertions (`assert param != null`). They break evaluation.

File: pkgs/applications/misc/k2pdfopt/default.nix [hunk 1]
```diff
@@ -28,6 +28,7 @@ stdenv.mkDerivation rec {
   let
     mupdf_modded = mupdf.overrideAttrs (attrs: {
       name = "mupdf-1.10a";
```
Comment (zimbatm) [inline pkgs/applications/misc/k2pdfopt/default.nix]:
```suggestion
      pname = "mupdf";
```

File: pkgs/applications/misc/oneko/default.nix [hunk 2]
```diff
@@ -1,11 +1,11 @@
 { stdenv, fetchurl, xorg, xlibsWrapper }:
 
 stdenv.mkDerivation rec {
-  version = "1.2.sakura.5";
-  vname = "1.2.5";
-  name = "oneko-${vname}";
+  vname = "1.2.sakura.5";
```
Comment (zimbatm) [inline pkgs/applications/misc/oneko/default.nix]:
```suggestion
  version_name = "1.2.sakura.5";
```

File: pkgs/applications/misc/oneko/default.nix [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
+  version = "1.2.5";
+  name = "oneko-${version}";
   src = fetchurl {
-    url = "http://www.daidouji.com/oneko/distfiles/oneko-${version}.tar.gz";
+    url = "http://www.daidouji.com/oneko/distfiles/oneko-${vname}.tar.gz";
```
Comment (zimbatm) [inline pkgs/applications/misc/oneko/default.nix]:
```suggestion
    url = "http://www.daidouji.com/oneko/distfiles/oneko-${version_name}.tar.gz";
```

File: pkgs/applications/networking/remote/remmina/default.nix [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 with stdenv.lib;
 
-stdenv.mkDerivation {
+stdenv.mkDerivation rec {
   name = "remmina-${version}";
```
Comment (zimbatm) [inline pkgs/applications/networking/remote/remmina/default.nix]:
```suggestion
  pname = "remmina";
```

File: pkgs/applications/virtualization/docker/default.nix [hunk 5]
```diff
@@ -28,7 +29,8 @@ rec {
     });
 
     docker-containerd = (containerd.override { inherit go; }).overrideAttrs (oldAttrs: rec {
-      name = "docker-containerd";
+      name = "docker-containerd-${version}";
```
Comment (zimbatm) [inline pkgs/applications/virtualization/docker/default.nix]: does this mean that the abstraction breaks down on override?
```suggestion
      pname = "docker-containerd";
```

File: pkgs/applications/virtualization/docker/default.nix [hunk 6]
```diff
@@ -42,7 +44,8 @@ rec {
     });
 
     docker-tini = tini.overrideAttrs  (oldAttrs: rec {
-      name = "docker-init";
+      name = "docker-init-${version}";
```
Comment (zimbatm) [inline pkgs/applications/virtualization/docker/default.nix]:
```suggestion
      pname = "docker-init";
```

**Review by zimbatm — APPROVED** [review]
if all of this compiles, then let's get this merged in first

File: pkgs/applications/misc/k2pdfopt/default.nix (hunk 1 above)
Comment (Synthetica9) [reply pkgs/applications/misc/k2pdfopt/default.nix]: Doesn't work, because the `name` attribute is already determined at this point

**Comment by vbgl** [comment]
Can we ensure that `mkDerivation` is lazy? We rely on this property on `coqPackages`.

**Comment by Synthetica9** [comment]
> Can we ensure that mkDerivation is lazy? We rely on this property in coqPackages.

Only in the places altered in 0ab59bc37b21a13dc35b9d902d78f26af4bb56f2? Because I am currently testing a fix that should fix the issues caused by that. It seems to work fine (testing by compiling coqPackages_8_8)

```
  param = params."${coq.coq-version}" or null;
in

if param == null then null else
stdenv.mkDerivation rec {
```

**Comment by vbgl** [comment]
What would be the value of `builtins.hasAttr "coqprime" coqPackages_8_5` with your fix? I expect “false”.

**Comment by Synthetica9** [comment]
@vbgl Everything should be okay now!

**Comment by vbgl** [comment]
Thanks for your understanding. LGTM.

**Comment by Synthetica9** [comment]
(Rebased to squash some commits, clarify a few messages, and solve merge conflicts)

File: pkgs/applications/networking/flent/http-getter.nix [hunk 7]
```diff
@@ -2,8 +2,8 @@
 , curl, pkgconfig }:
```
[the remark this answers is not in the document]
Comment (Synthetica9) [reply pkgs/applications/networking/flent/http-getter.nix]: [Yessir!](https://nixos.org/nixpkgs/manual/#sec-package-naming)

>  If a package is not a release but a commit from a repository, then the version part of the name must be the date of that (fetched) commit. The date must be in "YYYY-MM-DD" format. Also append "unstable" to the name - e.g., "pkgname-unstable-2014-09-23".

**Comment by Synthetica9** [comment]
Currently running on my Hydra (private, so no link, sorry!)

**Comment by Synthetica9** [comment]
@Mic92 Willing to give this another try?

**Comment by Synthetica9** [comment]
Okay, next order of business is doing refractors in the tree to make packages use this pattern. I think that should go in master, once this PR gets there, am I correct?

**Comment by timokau** [comment]
Usually. If it actually changes the resulting `name` it will cause rebuilds though. If the package is high enough in the dpendency tree, that should go to staging. So if ofBorg reports excessive (>100) rebuilds, rebase on staging.

**Comment by Synthetica9** [comment]
I was thinking of packages that define `name = "${pname}-${version}"`, which shouldn't cause rebuilds at all, right?

**Comment by timokau** [comment]
Not sure, since the default builder will add `pname` as an environment variable for the build I think.

**Comment by dtzWill** [comment]
`nixos-version` was failing this check for me--but I might have missed some of the new goodness, and due to strange network problems it's not easy to confidently check against master right now.  It's probably been fixed but if not thought I'd mention just in case :).

**Comment by Synthetica9** [comment]
See https://github.com/nlewo/nixpkgs/pull/1. Can't open a pr to staging atm
(laptop's empty), but feel free to cherry-pick

**Pull request merged**